### PR TITLE
Karim Wallani's submission for pre-employment assessment

### PR DESCRIPTION
**Pull request by karimw786**
This is Karim Wallani's pre-emloyment assessment submission. It contains all six tasks in separate commits, plus an additional commit to fix bugs and harden security.

### CODE REVIEW

File: src/Todo.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    function __construct($user_id, $description)
+    {
+        $this->user_id = (int) $user_id;
+        $this->description = (string) $description;
+    }
+
+    function save()
+    {
+        $sql = "INSERT INTO todos (user_id, description) VALUES ($this->user_id, ?)";
+        $result = $GLOBALS['app']['db']->executeUpdate($sql, array((string) $this->description));
```
Comment (T0RAT0RA) [inline src/Todo.php]: Avoid using `$GLOBALS` if possible.
You could inject it in the class constructor.

File: src/controllers.php [hunk 2]
```diff
@@ -10,80 +10,241 @@
 }));
 
 
-$app->get('/', function () use ($app) {
+$app->match('/', function () use ($app) {
```
Comment (T0RAT0RA) [inline src/controllers.php]: Why the change to `match` here?

File: src/controllers.php [hunk 3]
```diff
@@ -10,80 +10,241 @@
...
+            $app['session']->getFlashBag()->add('error', 'Username and password cannot be empty.');
         }
     }
 
     return $app['twig']->render('login.html', array());
 });
 
 
-$app->get('/logout', function () use ($app) {
+$app->match('/logout', function () use ($app) {
```
Comment (T0RAT0RA) [inline src/controllers.php]: Why the change to `match` here?

File: src/controllers.php [hunk 4]
```diff
@@ -10,80 +10,241 @@
...
 
-    if ($id){
-        $sql = "SELECT * FROM todos WHERE id = '$id'";
-        $todo = $app['db']->fetchAssoc($sql);
+    // Task 6
+    $todos = Todo::getAll($user['id']);
+
+    // Task 5
+    $totalItems = count($todos);
+    $pagination = new Pagination($totalItems, $page, 10);
```
Comment (T0RAT0RA) [inline src/controllers.php]: This is not the right way to do pagination.
You're always loading all todos for a user, what if he has 100 000 todos?

File: src/controllers.php [hunk 5]
```diff
@@ -10,80 +10,241 @@
...
+    ->convert(
+        'page',
+        function ($page) {
+            return (int) $page;
+        }
+    );
+
+
+$app->match('/todo/delete/{id}', function ($id, Request $request) use ($app) {
+    if($request->getMethod() == 'GET') {
```
Comment (T0RAT0RA) [inline src/controllers.php]: Not needed if you use `->post` instead of `->match`

File: src/controllers.php [hunk 6]
```diff
@@ -10,80 +10,241 @@
...
+    $pages = $pagination->build();
+    $currPage = $pagination->currentPage();
+    $prevPage = $currPage-1;
+
+    if ($currPage==0||$prevPage==0) {
+        $currPage=1;
+        $prevPage=1;
+    }
+
+    $app['session']->set('currentPage', $currPage);
```
Comment (T0RAT0RA) [inline src/controllers.php]: I'm curious why you store the pagination in session?

File: src/Todo.php (hunk 1 above)
Comment (karimw786) [reply src/Todo.php]: Oh OK, so you're saying that I should use $GLOBALS only once in the constructor?

File: src/controllers.php (hunk 2 above)
Comment (karimw786) [reply src/controllers.php]: I wanted to use get only when the function is taking information from the URL. In the case of "/", it's not really collecting anything from the URL, so I used match.

File: src/controllers.php (hunk 3 above)
Comment (karimw786) [reply src/controllers.php]: I wanted to use get only when the function is taking information from the URL. In the case of "/logout", it's not really collecting anything from the URL, so I used match.

File: src/controllers.php (hunk 4 above)
Comment (karimw786) [reply src/controllers.php]: That's a good point. I only did it this way because the Pagination Service Provider I was using suggested doing it this way in their examples. But I can see that doing it this way would become very costly when there is a large # of todos.

File: src/controllers.php (hunk 6 above)
Comment (karimw786) [reply src/controllers.php]: You're right, I should have kept currentPage, previousPage, totalItems, and itemsPerPage outside of any functions in controller.php, because I am only using these variables in controller.php, not any other script.

File: src/controllers.php (hunk 5 above)
Comment (karimw786) [reply src/controllers.php]: I had that before, but it was conflicting with ->get('/todo/{id}/{format}')
